sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,378 @@
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return (self.name == other.name and
+            self.genotype == other.genotype and
+            self.phaseset == other.phaseset and
+            self.info == other.info)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(filebasedsource.FileBasedSource):
+  r"""A source for reading VCF files.
+
+  Parses VCF files (version 4) using PyVCF library. If file_pattern specifies
+  multiple files, then the header from each file is used separately to parse
+  the content. However, the output will be a uniform PCollection of
+  :class:`Variant` objects.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64kB
+
+  def __init__(self,
+               file_pattern,
+               min_bundle_size=0,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern, min_bundle_size,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._coder = coders.StrUtf8Coder() # VCF files are UTF-8 per schema.
+    self._header_lines_per_file = {}
+    self._text_source = TextSource(
+        file_pattern,
+        min_bundle_size,
+        compression_type,
+        strip_trailing_newlines=True,
+        coder=self._coder,
+        buffer_size=buffer_size,
+        validate=validate,
+        skip_header_lines=0,
+        header_matcher_predicate=lambda x: x.startswith('#'))
+
+  def _process_header_lines_matching_predicate(self, file_name, header_lines):
+    self._header_lines_per_file[file_name] = header_lines
+
+  def read_records(self, file_name, range_tracker):
+    def line_generator():
+      header_processed = False
+      for line in self._text_source.read_records(
+          file_name,
+          range_tracker,
+          self._process_header_lines_matching_predicate):
+        if not header_processed and file_name in self._header_lines_per_file:
+          for header in self._header_lines_per_file[file_name]:
+            yield header
+          header_processed = True
+        # PyVCF has explicit str() calls when parsing INFO fields, which fails
+        # with UTF-8 decoded strings. Encode the line back to UTF-8.
+        yield self._coder.encode(line)
+    try:
+      vcf_reader = vcf.Reader(fsock=line_generator())
+    except SyntaxError as e:
+      raise ValueError('Invalid VCF header: %s' % str(e))
+    while True:
+      try:
+        record = next(vcf_reader)
+        yield self._convert_to_variant_record(
+            record, vcf_reader.infos, vcf_reader.formats)
+      except StopIteration:
+        break
+      except (LookupError, ValueError) as e:
```

Comment:
These seem like extremely broad classes of errors to catch (pretty much any bug in the code would cause one of these errors). What's the benefit of catching them at all, rather than letting them propagate?

Reply:
@jkff: The reason for this is that pyVCF throws these very generic errors. Currently, we want execution to continue if an invalid VCF file is used so that other valid VCF files can still be processed. pyVCF doesn't do any error checking to see if the VCF file is valid so it throws a lookup error in cases like this: https://github.com/jamescasbon/PyVCF/blob/master/vcf/parser.py#L555

Reply:
Hm, not sure I'm following:

* The current does does _not_ let execution continue after the error - it raises a ValueError wrapping the original error
* After an error, letting execution continue would not be a great idea, see https://beam.apache.org/contribute/ptransform-style-guide/#runtime-errors-and-data-consistency . It is better to represent errors explicitly as the elements of your transform's result. For example, it could yield objects that are either VcfRecord or a marker MalformedVcfFile(filename, line) that this file was truncated (assuming it's possible to isolate the error to an individual line that failed).

Reply:
@jkff: Sorry, I was mistaken, execution does halt when an error occurs. The current implementation does what we call 'strict' mode where a failure to properly read a single record causes a halt in execution. We plan to eventually also implement a 'loose' mode that would handle exceptions in the way you described.